modules/pubxaiAnalyticsAdapter.js

```diff
@@ -13,196 +23,335 @@ const defaultHost = 'api.pbxai.com';
-      mediaType: bidResponse.mediaType,
-      netRevenue: bidResponse.netRevenue,
-      requestTimestamp: bidResponse.requestTimestamp,
-      responseTimestamp: bidResponse.responseTimestamp,
-      status: bidResponse.status,
-      statusMessage: bidResponse.statusMessage,
-      timeToRespond: bidResponse.timeToRespond,
-      transactionId: bidResponse.transactionId
-    };
-    if (status !== 'bidwon') {
-      Object.assign(bid, {
-        bidId: status === 'timeout' ? bidResponse.bidId : bidResponse.requestId,
-        renderStatus: status === 'timeout' ? 3 : 2,
-        sizes: parseSizesInput(bidResponse.size).toString(),
+/**
+ * Track the events emitted by prebid and handle each case. See https://docs.prebid.org/dev-docs/publisher-api-reference/getEvents.html for more info
+ * @param {object} event the prebid event emmitted
+ * @param {string} event.eventType the type of the event
+ * @param {object} event.args the arguments of the emitted event
+ */
+const track = ({ eventType, args }) => {
+  switch (eventType) {
+    // handle invalid bids, and remove them from the adUnit cache
+    case CONSTANTS.EVENTS.BID_TIMEOUT:
+      args.map(extractBid).forEach((bid) => {
+        bid.renderStatus = 3;
+        auctionCache[bid.auctionId].bids.push(bid);
+      });
+      break;
+    // handle valid bid responses and record them as part of an auction
+    case CONSTANTS.EVENTS.BID_RESPONSE:
+      const bid = Object.assign(extractBid(args), { renderStatus: 2 });
+      auctionCache[bid.auctionId].bids.push(bid);
+      break;
+    // capture extra information from the auction, and if there were no bids
+    // (and so no chance of a win) send the auction
+    case CONSTANTS.EVENTS.AUCTION_END:
+      Object.assign(
+        auctionCache[args.auctionId].floorDetail,
+        deepAccess(args, 'adUnits.0.bids.0.floorData')
+      );
+      auctionCache[args.auctionId].pageDetail.adUnits = args.adUnitCodes;
+      Object.assign(auctionCache[args.auctionId].auctionDetail, {
+        adUnitCodes: args.adUnits.map((i) => i.code),
+        timestamp: args.timestamp,
       });
-      events.bids.push(bid);
-    } else {
-      Object.assign(bid, {
-        bidId: bidResponse.requestId,
-        floorProvider: events.floorDetail?.floorProvider || null,
-        floorFetchStatus: events.floorDetail?.fetchStatus || null,
-        floorLocation: events.floorDetail?.location || null,
-        floorModelVersion: events.floorDetail?.modelVersion || null,
-        floorSkipRate: events.floorDetail?.skipRate || 0,
-        isFloorSkipped: events.floorDetail?.skipped || false,
+      if (
+        auctionCache[args.auctionId].bids.every((bid) => bid.renderStatus === 3)
+      ) {
+        send(args.auctionId);
+      }
+      break;
+    // send the prebid winning bid back to pubx
+    case CONSTANTS.EVENTS.BID_WON:
+      const winningBid = extractBid(args);
+      const floorDetail = auctionCache[winningBid.auctionId].floorDetail;
```

Comment:
this can happen before auctionEnd as well, so better to get it the same way we get it in auctionEnd or save it during the auctionInit itself.

Reply:
After double checking the code, it is _technically_ possible for a publisher to operate such that they render the first bid that is returned by prebid, without waiting for any other bids to be returned, or for the end of the auction. This would be a publisher error, and would take quite a substantial amount of work for a publisher to do. (If you look at src/Prebid.js, src/auction.js and src/adRendering.js, you'll see that before prebid will render an ad, it expects to have accepted a suitable bid request for the ad slot, which in turn is managed by the auction_init event and registered as a callback in the running auction itself. As far as i can tell, there are no hooks available to "cancel" the running auction and interject a winning bid callback before the auction has ended, without implicitly ending the auction as a side effect of running the bidwon callback.)

Without operating in a malicious way, this case is "impossible", and so after discussion with PHN i've left this logic as-is.